File: 1004.max-consecutive-ones-iii.cpp
/*
 * @lc app=leetcode id=1004 lang=cpp
 *
 * [1004] Max Consecutive Ones III
 */
#include <vector>

using std::vector;

// @lc code=start
class Solution {
public:
    // Sliding window:
    //     If A[l:r+1] has zeros <= K, we continue to increment r (expand window to the right).
    //     If A[l:r+1] has zeros > K, we increment l (shift the left boundary) as well as r.
    // T: O(n)
    // S: O(1)
    int longestOnes(vector<int>& nums, int k) {
        int l = 0;
        int r = 0;
        for (r = 0; r < nums.size(); ++r) {
            if (nums[r] == 0) {
                --k;
            }
            if (k < 0 && nums[l++] == 0) {
                // Too many zeros to be flipped,
                // shift the left window boundary
                ++k;
            }
        }
        return r - l;
    }
};
// @lc code=end

File: 1004.max-consecutive-ones-iii.py
#
# @lc app=leetcode id=1004 lang=python3
#
# [1004] Max Consecutive Ones III
#
from typing import List

# @lc code=start
class Solution:
    # Sliding window
    # T: O(n)
    # S: O(1)
    def longestOnes(self, nums: List[int], k: int) -> int:
        l = r = 0
        for r in range(len(nums)):
            # Try extending the window to the right,
            # and update the number of zeros.
            if not nums[r]:
                k -= 1

            if k < 0:
                # More than k zeros inside the window.
                # Try shifting the left boundary to
                # reduce the number of zeros.
                if not nums[l]:
                    k += 1
                l += 1
        return r - l + 1

# @lc code=end

